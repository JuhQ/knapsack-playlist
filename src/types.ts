export interface YoutubeItem {
  id: string;
  title: string;
  seconds: number;
  rating: number;
}
