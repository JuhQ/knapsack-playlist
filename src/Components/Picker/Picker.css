.picker-container {
  padding-top: 0.125rem;
}

.picker-submit {
  margin-top: 1rem;
  margin-bottom: 1rem;
}

.picker-container .ui.items > .item div,
.picker-container .ui.items > .item > .content > .header:not(.ui) {
  color: white;
}
