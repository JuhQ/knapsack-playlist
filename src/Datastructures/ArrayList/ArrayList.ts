/**
 * ArrayList implementation in typescript
 */
class ArrayList<T> {
  private list: T[];

  private index: number;

  constructor(size: number) {
    this.list = Array<T>(size);
    this.index = 0;
  }

  // Time complexity O(1)
  private isFull(): boolean {
    return this.index + 1 === this.list.length;
  }

  // Time complexity O(n)
  // Space complexity O(n)
  private grow(): T[] {
    // reserve three times as much space for the next list, so we don't need to do this operation too often
    const list = new Array<T>(this.size() * 3);

    for (let i = 0; i < this.size(); i++) {
      list[i] = this.list[i];
    }

    return list;
  }

  // Time complexity O(1), worst case O(n)
  push(value: T): void {
    if (this.isFull()) {
      this.list = this.grow();
    }

    this.list[this.index] = value;
    this.index++;
  }

  // Time complexity O(n)
  // Space complexity O(n)
  slice(start?: number, end?: number): ArrayList<T> {
    const sliced = this.getAsArray().slice(start, end);

    const list = new ArrayList<T>(sliced.length);

    for (let i = 0; i < sliced.length; i++) {
      list.push(sliced[i]);
    }

    return list;
  }

  // remove item from the beginning of the queue and return that value
  // Time complexity O(n)
  shift(): T | undefined {
    this.index--;
    return this.list.shift();
  }

  // Time complexity O(n)
  // Space complexity O(n)
  map<K>(fn: (value: T) => K): ArrayList<K> {
    const list = this.getAsArray();
    const newList = new ArrayList<K>(this.size());

    for (let i = 0; i < list.length; i++) {
      newList.push(fn(list[i]));
    }

    return newList;
  }

  // Time complexity O(n)
  // Space complexity O(n)
  filter(fn: (value: T) => boolean): ArrayList<T> {
    const list = this.getAsArray();
    const newList = new ArrayList<T>(1);

    for (let i = 0; i < list.length; i++) {
      if (fn(list[i])) {
        newList.push(list[i]);
      }
    }

    return newList;
  }

  // Time complexity O(n)
  // Space complexity O(n)
  find<K>(fn: (value: T) => K): T | undefined {
    const list = this.getAsArray();

    for (let i = 0; i < list.length; i++) {
      if (fn(list[i])) {
        return list[i];
      }
    }

    return undefined;
  }

  // TODO: implement proper sort algorithm
  sort(fn: (a: T, b: T) => number): ArrayList<T> {
    const sorted = [...this.list].sort(fn);

    const list = new ArrayList<T>(this.size());
    for (let i = 0; i < this.size(); i++) {
      list.push(sorted[i]);
    }

    return list;
  }

  // Time complexity O(1)
  size(): number {
    return this.index;
  }

  // Time complexity O(1)
  at(index: number): T {
    return this.list[index];
  }

  /**
   * Return all values from the list, except undefined values
   * This will cause issue if given type would be undefined, but in this project undefined lists will never be used
   * Undefined is removed because the grow method will reserve extra space
   * Time complexity O(n)
   * Space complexity O(n)
   */
  getAsArray(): T[] {
    const list: T[] = [];
    for (let i = 0; i < this.list.length; i++) {
      if (this.list[i] !== undefined) {
        list.push(this.list[i]);
      }
    }

    return list;
  }
}

export default ArrayList;
