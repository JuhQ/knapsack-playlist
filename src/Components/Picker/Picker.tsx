import "./Picker.css";

import React, { useState } from "react";
import { Input, Item } from "semantic-ui-react";

import { YoutubeItem } from "../../types";
import { sumSeconds } from "../../Utils/math";
import PickerItem from "./PickerItem/PickerItem";

interface SubmitReturn {
  list: YoutubeItem[];
  length: number;
}

interface Props {
  list: YoutubeItem[];
  onSubmit: ({ list }: SubmitReturn) => void;
}

const Picker: React.FC<Props> = ({ list, onSubmit }: Props) => {
  const [playlistLength, setPlaylistLength] = useState<number>(3600);
  const [search, setSearch] = useState<string>("");
  const [selectedList, setSelectedList] = useState<YoutubeItem[]>([]);

  const handleSelection = (item: YoutubeItem) => {
    const itemFound = selectedList.find(({ id }) => item.id === id);

    if (itemFound) {
      setSelectedList(selectedList.filter(({ id }) => id !== item.id));
    } else {
      setSelectedList([...selectedList, item]);
    }
  };

  const length = sumSeconds(selectedList);

  return (
    <form
      onSubmit={(event) => {
        event.preventDefault();
        onSubmit({ list: selectedList, length: playlistLength });
      }}
    >
      <h3>Define desired playlist length and pick your music</h3>
      <Input
        size="large"
        type="number"
        label={{ basic: true, content: "seconds" }}
        labelPosition="right"
        placeholder="Enter desired playlist length"
        value={playlistLength}
        onChange={({ target }) => {
          setPlaylistLength(Number(target.value));
        }}
      />
      <br />
      <button type="submit" className="picker-submit">
        Click here to generate the playlist from music you picked
      </button>
      <br />
      <small>
        Depending on the size of the set you pick, and the length of the
        playlist you set, the generated playlist may or may not include all the
        songs you picked.
      </small>
      <br />
      List duration {length}
      <br />
      List size {selectedList.length}
      <br />
      <Input
        type="search"
        placeholder="Search"
        onChange={({ target }) => {
          setSearch(target.value);
        }}
      />
      <div className="picker-container">
        <Item.Group>
          {list
            .filter((item) =>
              search.length
                ? item.title.toLowerCase().includes(search.toLowerCase())
                : true
            )
            .map((item) => (
              <PickerItem
                item={item}
                onClick={() => handleSelection(item)}
                key={item.id}
              />
            ))}
        </Item.Group>
      </div>
    </form>
  );
};

export default Picker;
