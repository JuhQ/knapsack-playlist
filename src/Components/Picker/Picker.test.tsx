import { shallow } from "enzyme";
import { shallowToJson } from "enzyme-to-json";
import React from "react";
import renderer, { act } from "react-test-renderer";

import ArrayList from "../../Datastructures/ArrayList/ArrayList";
import Queue from "../../Datastructures/queue/queue";
import { YoutubeItem } from "../../types";
import generateTestData from "../../Utils/testHelpers";
import Picker from "./Picker";

interface PaginationProps {
  prop: (
    value: string
  ) => (
    event: { preventDefault: () => null },
    result: { activePage: number }
  ) => null;
}

const event = { preventDefault: jest.fn() };

describe("Picker", () => {
  beforeEach(() => {
    jest.spyOn(event, "preventDefault");
  });

  it("should render", () => {
    const list = new ArrayList<YoutubeItem>(2);
    list.push({ id: "1", title: "test", seconds: 60, rating: 1 });
    list.push({ id: "2", title: "test 2", seconds: 60, rating: 1 });
    const queue = new Queue();
    queue.merge(list);

    const component = renderer.create(
      <Picker list={queue} onSubmit={() => null} />
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("should be able to call onSubmit function", () => {
    const list = new ArrayList<YoutubeItem>(2);
    list.push({ id: "1", title: "test", seconds: 60, rating: 1 });
    list.push({ id: "2", title: "test 2", seconds: 60, rating: 1 });
    const queue = new Queue();
    queue.merge(list);

    const handleSubmit = jest.fn();
    const wrapper = shallow(<Picker list={queue} onSubmit={handleSubmit} />);
    expect(handleSubmit).not.toHaveBeenCalled();
    wrapper.simulate("submit", event);
    expect(event.preventDefault).toBeCalled();
    expect(handleSubmit).toHaveBeenCalled();
  });

  it("should be able to call onSubmit function and return a list of selected songs", () => {
    const list = new ArrayList<YoutubeItem>(2);
    list.push({ id: "1", title: "test", seconds: 60, rating: 1 });
    list.push({ id: "2", title: "test 2", seconds: 60, rating: 1 });
    const queue = new Queue();
    queue.merge(list);

    const handleSubmit = jest.fn();
    const wrapper = shallow(<Picker list={queue} onSubmit={handleSubmit} />);
    expect(handleSubmit).not.toHaveBeenCalled();
    wrapper.find("PickerItem").last().simulate("click");
    wrapper.simulate("submit", event);
    expect(event.preventDefault).toBeCalled();
    expect(handleSubmit).toHaveBeenCalledWith({
      length: 3600,
      list: {
        enqueue: expect.any(Function),
        list: {
          index: 1,
          list: [{ id: "2", seconds: 60, title: "test 2", rating: 1 }],
        },
      },
    });

    act(() => {
      expect(shallowToJson(wrapper)).toMatchSnapshot();
    });
  });

  it("should be able to call filter list", () => {
    const list = new ArrayList<YoutubeItem>(2);
    list.push({ id: "1", title: "test", seconds: 60, rating: 1 });
    list.push({ id: "2", title: "test 2", seconds: 60, rating: 1 });
    const queue = new Queue();
    queue.merge(list);

    const handleSubmit = jest.fn();
    const wrapper = shallow(<Picker list={queue} onSubmit={handleSubmit} />);
    expect(handleSubmit).not.toHaveBeenCalled();
    wrapper
      .find("Input")
      .last()
      .simulate("change", { target: { value: "rammstein" } });

    act(() => {
      expect(shallowToJson(wrapper)).toMatchSnapshot();
    });
  });

  it("should be able to call onSubmit function and return an empty list because selection was removed on the second click", () => {
    const list = new ArrayList<YoutubeItem>(2);
    list.push({ id: "1", title: "test", seconds: 60, rating: 1 });
    list.push({ id: "2", title: "test 2", seconds: 60, rating: 1 });
    const queue = new Queue();
    queue.merge(list);

    const handleSubmit = jest.fn();
    const wrapper = shallow(<Picker list={queue} onSubmit={handleSubmit} />);

    expect(handleSubmit).not.toHaveBeenCalled();
    wrapper.find("PickerItem").last().simulate("click");
    wrapper.find("PickerItem").last().simulate("click");
    wrapper.simulate("submit", event);
    expect(event.preventDefault).toBeCalled();
    expect(handleSubmit).toHaveBeenCalledWith({
      length: 3600,
      list: {
        enqueue: expect.any(Function),
        list: {
          index: 0,
          list: [undefined],
        },
      },
    });

    act(() => {
      expect(shallowToJson(wrapper)).toMatchSnapshot();
    });
  });

  it("should be able to submit with length", () => {
    const list = new ArrayList<YoutubeItem>(2);
    list.push({ id: "1", title: "test", seconds: 60, rating: 1 });
    list.push({ id: "2", title: "test 2", seconds: 60, rating: 1 });
    const queue = new Queue();
    queue.merge(list);

    const handleSubmit = jest.fn();
    const wrapper = shallow(<Picker list={queue} onSubmit={handleSubmit} />);
    expect(handleSubmit).not.toHaveBeenCalled();

    wrapper
      .find("Input")
      .first()
      .simulate("change", { target: { value: 200 } });
    wrapper.find("PickerItem").last().simulate("click");

    wrapper.simulate("submit", event);

    expect(handleSubmit).toHaveBeenCalled();
    expect(handleSubmit).toHaveBeenCalledWith({
      length: 200,
      list: {
        enqueue: expect.any(Function),
        list: {
          index: 1,
          list: [{ id: "2", seconds: 60, title: "test 2", rating: 1 }],
        },
      },
    });
  });

  it("should render multiple items with pagination", () => {
    const wrapper = shallow(
      <Picker list={generateTestData(30, 20)} onSubmit={() => null} />
    );

    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });

  it("should be able change page", () => {
    const wrapper = shallow(
      <Picker list={generateTestData(30, 20)} onSubmit={() => null} />
    );

    act(() => {
      // simulate does not work here :(
      const pagination: PaginationProps = wrapper.find("Pagination");
      pagination.prop("onPageChange")(event, { activePage: 2 });
      expect(shallowToJson(wrapper)).toMatchSnapshot();
    });
  });
});
