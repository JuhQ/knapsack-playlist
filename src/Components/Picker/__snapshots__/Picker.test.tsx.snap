// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Picker should be able change page 1`] = `
<form
  onSubmit={[Function]}
>
  <h3>
    Define desired playlist length and pick your music
  </h3>
  <Input
    label={
      Object {
        "basic": true,
        "content": "seconds",
      }
    }
    labelPosition="right"
    onChange={[Function]}
    placeholder="Enter desired playlist length"
    size="large"
    type="number"
    value={3600}
  />
  <br />
  <button
    className="picker-submit"
    type="submit"
  >
    Click here to generate the playlist from music you picked
  </button>
  <br />
  <small>
    Depending on the size of the set you pick, and the length of the playlist you set, the generated playlist may or may not include all the songs you picked.
  </small>
  <br />
  List duration 
  0
  <br />
  List size 
  0
  <br />
  <Input
    onChange={[Function]}
    placeholder="Search"
    type="search"
  />
  <div
    className="picker-container"
  >
    <Pagination
      activePage={2}
      aria-label="Pagination Navigation"
      boundaryRange={1}
      ellipsisItem="..."
      firstItem={
        Object {
          "aria-label": "First item",
          "content": "«",
        }
      }
      lastItem={
        Object {
          "aria-label": "Last item",
          "content": "»",
        }
      }
      nextItem={
        Object {
          "aria-label": "Next item",
          "content": "⟩",
        }
      }
      onPageChange={[Function]}
      pageItem={Object {}}
      prevItem={
        Object {
          "aria-label": "Previous item",
          "content": "⟨",
        }
      }
      siblingRange={1}
      totalPages={3}
    />
    <ItemGroup>
      <PickerItem
        item={
          Object {
            "id": "10",
            "seconds": 20,
            "title": "test",
          }
        }
        key="10"
        onClick={[Function]}
      />
      <PickerItem
        item={
          Object {
            "id": "11",
            "seconds": 20,
            "title": "test",
          }
        }
        key="11"
        onClick={[Function]}
      />
      <PickerItem
        item={
          Object {
            "id": "12",
            "seconds": 20,
            "title": "test",
          }
        }
        key="12"
        onClick={[Function]}
      />
      <PickerItem
        item={
          Object {
            "id": "13",
            "seconds": 20,
            "title": "test",
          }
        }
        key="13"
        onClick={[Function]}
      />
      <PickerItem
        item={
          Object {
            "id": "14",
            "seconds": 20,
            "title": "test",
          }
        }
        key="14"
        onClick={[Function]}
      />
      <PickerItem
        item={
          Object {
            "id": "15",
            "seconds": 20,
            "title": "test",
          }
        }
        key="15"
        onClick={[Function]}
      />
      <PickerItem
        item={
          Object {
            "id": "16",
            "seconds": 20,
            "title": "test",
          }
        }
        key="16"
        onClick={[Function]}
      />
      <PickerItem
        item={
          Object {
            "id": "17",
            "seconds": 20,
            "title": "test",
          }
        }
        key="17"
        onClick={[Function]}
      />
      <PickerItem
        item={
          Object {
            "id": "18",
            "seconds": 20,
            "title": "test",
          }
        }
        key="18"
        onClick={[Function]}
      />
      <PickerItem
        item={
          Object {
            "id": "19",
            "seconds": 20,
            "title": "test",
          }
        }
        key="19"
        onClick={[Function]}
      />
    </ItemGroup>
  </div>
</form>
`;

exports[`Picker should be able to call filter list 1`] = `
<form
  onSubmit={[Function]}
>
  <h3>
    Define desired playlist length and pick your music
  </h3>
  <Input
    label={
      Object {
        "basic": true,
        "content": "seconds",
      }
    }
    labelPosition="right"
    onChange={[Function]}
    placeholder="Enter desired playlist length"
    size="large"
    type="number"
    value={3600}
  />
  <br />
  <button
    className="picker-submit"
    type="submit"
  >
    Click here to generate the playlist from music you picked
  </button>
  <br />
  <small>
    Depending on the size of the set you pick, and the length of the playlist you set, the generated playlist may or may not include all the songs you picked.
  </small>
  <br />
  List duration 
  0
  <br />
  List size 
  0
  <br />
  <Input
    onChange={[Function]}
    placeholder="Search"
    type="search"
  />
  <div
    className="picker-container"
  >
    <Pagination
      activePage={1}
      aria-label="Pagination Navigation"
      boundaryRange={1}
      ellipsisItem="..."
      firstItem={
        Object {
          "aria-label": "First item",
          "content": "«",
        }
      }
      lastItem={
        Object {
          "aria-label": "Last item",
          "content": "»",
        }
      }
      nextItem={
        Object {
          "aria-label": "Next item",
          "content": "⟩",
        }
      }
      onPageChange={[Function]}
      pageItem={Object {}}
      prevItem={
        Object {
          "aria-label": "Previous item",
          "content": "⟨",
        }
      }
      siblingRange={1}
      totalPages={0}
    />
    <ItemGroup />
  </div>
</form>
`;

exports[`Picker should be able to call onSubmit function and return a list of selected songs 1`] = `
<form
  onSubmit={[Function]}
>
  <h3>
    Define desired playlist length and pick your music
  </h3>
  <Input
    label={
      Object {
        "basic": true,
        "content": "seconds",
      }
    }
    labelPosition="right"
    onChange={[Function]}
    placeholder="Enter desired playlist length"
    size="large"
    type="number"
    value={3600}
  />
  <br />
  <button
    className="picker-submit"
    type="submit"
  >
    Click here to generate the playlist from music you picked
  </button>
  <br />
  <small>
    Depending on the size of the set you pick, and the length of the playlist you set, the generated playlist may or may not include all the songs you picked.
  </small>
  <br />
  List duration 
  60
  <br />
  List size 
  1
  <br />
  <Input
    onChange={[Function]}
    placeholder="Search"
    type="search"
  />
  <div
    className="picker-container"
  >
    <Pagination
      activePage={1}
      aria-label="Pagination Navigation"
      boundaryRange={1}
      ellipsisItem="..."
      firstItem={
        Object {
          "aria-label": "First item",
          "content": "«",
        }
      }
      lastItem={
        Object {
          "aria-label": "Last item",
          "content": "»",
        }
      }
      nextItem={
        Object {
          "aria-label": "Next item",
          "content": "⟩",
        }
      }
      onPageChange={[Function]}
      pageItem={Object {}}
      prevItem={
        Object {
          "aria-label": "Previous item",
          "content": "⟨",
        }
      }
      siblingRange={1}
      totalPages={1}
    />
    <ItemGroup>
      <PickerItem
        item={
          Object {
            "id": "1",
            "rating": 1,
            "seconds": 60,
            "title": "test",
          }
        }
        key="1"
        onClick={[Function]}
      />
      <PickerItem
        item={
          Object {
            "id": "2",
            "rating": 1,
            "seconds": 60,
            "title": "test 2",
          }
        }
        key="2"
        onClick={[Function]}
      />
    </ItemGroup>
  </div>
</form>
`;

exports[`Picker should be able to call onSubmit function and return an empty list because selection was removed on the second click 1`] = `
<form
  onSubmit={[Function]}
>
  <h3>
    Define desired playlist length and pick your music
  </h3>
  <Input
    label={
      Object {
        "basic": true,
        "content": "seconds",
      }
    }
    labelPosition="right"
    onChange={[Function]}
    placeholder="Enter desired playlist length"
    size="large"
    type="number"
    value={3600}
  />
  <br />
  <button
    className="picker-submit"
    type="submit"
  >
    Click here to generate the playlist from music you picked
  </button>
  <br />
  <small>
    Depending on the size of the set you pick, and the length of the playlist you set, the generated playlist may or may not include all the songs you picked.
  </small>
  <br />
  List duration 
  0
  <br />
  List size 
  0
  <br />
  <Input
    onChange={[Function]}
    placeholder="Search"
    type="search"
  />
  <div
    className="picker-container"
  >
    <Pagination
      activePage={1}
      aria-label="Pagination Navigation"
      boundaryRange={1}
      ellipsisItem="..."
      firstItem={
        Object {
          "aria-label": "First item",
          "content": "«",
        }
      }
      lastItem={
        Object {
          "aria-label": "Last item",
          "content": "»",
        }
      }
      nextItem={
        Object {
          "aria-label": "Next item",
          "content": "⟩",
        }
      }
      onPageChange={[Function]}
      pageItem={Object {}}
      prevItem={
        Object {
          "aria-label": "Previous item",
          "content": "⟨",
        }
      }
      siblingRange={1}
      totalPages={1}
    />
    <ItemGroup>
      <PickerItem
        item={
          Object {
            "id": "1",
            "rating": 1,
            "seconds": 60,
            "title": "test",
          }
        }
        key="1"
        onClick={[Function]}
      />
      <PickerItem
        item={
          Object {
            "id": "2",
            "rating": 1,
            "seconds": 60,
            "title": "test 2",
          }
        }
        key="2"
        onClick={[Function]}
      />
    </ItemGroup>
  </div>
</form>
`;

exports[`Picker should render 1`] = `
<form
  onSubmit={[Function]}
>
  <h3>
    Define desired playlist length and pick your music
  </h3>
  <div
    className="ui large right labeled input"
  >
    <input
      onChange={[Function]}
      placeholder="Enter desired playlist length"
      type="number"
      value={3600}
    />
    <div
      className="ui basic label label"
      onClick={[Function]}
    >
      seconds
    </div>
  </div>
  <br />
  <button
    className="picker-submit"
    type="submit"
  >
    Click here to generate the playlist from music you picked
  </button>
  <br />
  <small>
    Depending on the size of the set you pick, and the length of the playlist you set, the generated playlist may or may not include all the songs you picked.
  </small>
  <br />
  List duration 
  0
  <br />
  List size 
  0
  <br />
  <div
    className="ui input"
  >
    <input
      onChange={[Function]}
      placeholder="Search"
      type="search"
    />
  </div>
  <div
    className="picker-container"
  >
    <div
      aria-label="Pagination Navigation"
      className="ui pagination menu"
      role="navigation"
    >
      <a
        aria-current={false}
        aria-disabled={false}
        aria-label="First item"
        className="item"
        onClick={[Function]}
        onKeyDown={[Function]}
        tabIndex={0}
        type="firstItem"
        value={1}
      >
        «
      </a>
      <a
        aria-current={false}
        aria-disabled={false}
        aria-label="Previous item"
        className="item"
        onClick={[Function]}
        onKeyDown={[Function]}
        tabIndex={0}
        type="prevItem"
        value={1}
      >
        ⟨
      </a>
      <a
        aria-current={true}
        aria-disabled={false}
        className="active item"
        onClick={[Function]}
        onKeyDown={[Function]}
        tabIndex={0}
        type="pageItem"
        value={1}
      >
        1
      </a>
      <a
        aria-current={false}
        aria-disabled={false}
        aria-label="Next item"
        className="item"
        onClick={[Function]}
        onKeyDown={[Function]}
        tabIndex={0}
        type="nextItem"
        value={1}
      >
        ⟩
      </a>
      <a
        aria-current={false}
        aria-disabled={false}
        aria-label="Last item"
        className="item"
        onClick={[Function]}
        onKeyDown={[Function]}
        tabIndex={0}
        type="lastItem"
        value={1}
      >
        »
      </a>
    </div>
    <div
      className="ui items"
    >
      <a
        className="item"
        onClick={[Function]}
      >
        <div
          className="image"
        >
          <img
            src="https://i.ytimg.com/vi/1/hqdefault.jpg"
          />
        </div>
        <div
          className="content"
        >
          <div
            className="header"
          >
            test
          </div>
          <div
            className="meta"
          >
            1 minute
             
            
          </div>
        </div>
      </a>
      <a
        className="item"
        onClick={[Function]}
      >
        <div
          className="image"
        >
          <img
            src="https://i.ytimg.com/vi/2/hqdefault.jpg"
          />
        </div>
        <div
          className="content"
        >
          <div
            className="header"
          >
            test 2
          </div>
          <div
            className="meta"
          >
            1 minute
             
            
          </div>
        </div>
      </a>
    </div>
  </div>
</form>
`;

exports[`Picker should render multiple items with pagination 1`] = `
<form
  onSubmit={[Function]}
>
  <h3>
    Define desired playlist length and pick your music
  </h3>
  <Input
    label={
      Object {
        "basic": true,
        "content": "seconds",
      }
    }
    labelPosition="right"
    onChange={[Function]}
    placeholder="Enter desired playlist length"
    size="large"
    type="number"
    value={3600}
  />
  <br />
  <button
    className="picker-submit"
    type="submit"
  >
    Click here to generate the playlist from music you picked
  </button>
  <br />
  <small>
    Depending on the size of the set you pick, and the length of the playlist you set, the generated playlist may or may not include all the songs you picked.
  </small>
  <br />
  List duration 
  0
  <br />
  List size 
  0
  <br />
  <Input
    onChange={[Function]}
    placeholder="Search"
    type="search"
  />
  <div
    className="picker-container"
  >
    <Pagination
      activePage={1}
      aria-label="Pagination Navigation"
      boundaryRange={1}
      ellipsisItem="..."
      firstItem={
        Object {
          "aria-label": "First item",
          "content": "«",
        }
      }
      lastItem={
        Object {
          "aria-label": "Last item",
          "content": "»",
        }
      }
      nextItem={
        Object {
          "aria-label": "Next item",
          "content": "⟩",
        }
      }
      onPageChange={[Function]}
      pageItem={Object {}}
      prevItem={
        Object {
          "aria-label": "Previous item",
          "content": "⟨",
        }
      }
      siblingRange={1}
      totalPages={3}
    />
    <ItemGroup>
      <PickerItem
        item={
          Object {
            "id": "0",
            "seconds": 20,
            "title": "test",
          }
        }
        key="0"
        onClick={[Function]}
      />
      <PickerItem
        item={
          Object {
            "id": "1",
            "seconds": 20,
            "title": "test",
          }
        }
        key="1"
        onClick={[Function]}
      />
      <PickerItem
        item={
          Object {
            "id": "2",
            "seconds": 20,
            "title": "test",
          }
        }
        key="2"
        onClick={[Function]}
      />
      <PickerItem
        item={
          Object {
            "id": "3",
            "seconds": 20,
            "title": "test",
          }
        }
        key="3"
        onClick={[Function]}
      />
      <PickerItem
        item={
          Object {
            "id": "4",
            "seconds": 20,
            "title": "test",
          }
        }
        key="4"
        onClick={[Function]}
      />
      <PickerItem
        item={
          Object {
            "id": "5",
            "seconds": 20,
            "title": "test",
          }
        }
        key="5"
        onClick={[Function]}
      />
      <PickerItem
        item={
          Object {
            "id": "6",
            "seconds": 20,
            "title": "test",
          }
        }
        key="6"
        onClick={[Function]}
      />
      <PickerItem
        item={
          Object {
            "id": "7",
            "seconds": 20,
            "title": "test",
          }
        }
        key="7"
        onClick={[Function]}
      />
      <PickerItem
        item={
          Object {
            "id": "8",
            "seconds": 20,
            "title": "test",
          }
        }
        key="8"
        onClick={[Function]}
      />
      <PickerItem
        item={
          Object {
            "id": "9",
            "seconds": 20,
            "title": "test",
          }
        }
        key="9"
        onClick={[Function]}
      />
    </ItemGroup>
  </div>
</form>
`;
